## Operaciones básicas
La clase array de Numpy es extremadamente flexible y eso es en gran parte porque nos permite operaciones.

Las operaciones **suma** y **resta** requieren que los arrays tengan la misma forma, es decir, mismo número y tamaño de las dimensiones.

En el caso del **producto**, la **divisón** y la **potencia** se pueden operar arrays de las mismas dimensiones si el número de columnas de la primera coincide con el número de filas de la segunda.

### Suma

In [2]:
import numpy as np

# Dados dos arrays
arr_1 = np.array([1,2,3,4])
arr_2 = np.array([5,6,7,8])

# Los sumamos
arr_1 + arr_2

array([ 6,  8, 10, 12])

Si no tienen la misma forma no podemos sumarlos:

In [45]:
arr_3 = np.array([9,10])

arr_2 + arr_3

ValueError: operands could not be broadcast together with shapes (4,) (2,) 

### Resta

In [46]:
arr_2 - arr_1

array([4, 4, 4, 4])

¿Qué ocurriría si restamos un array a si mismo?

In [47]:
arr_1 - arr_1

array([0, 0, 0, 0])

Pues que obtenemos un array con todos los valores a cero.

### Producto

In [48]:
arr_1 * arr_2

array([ 5, 12, 21, 32])

El producto entre arrays se basa en multiplicar cada elemento de una array por el elemento en la misma posición del otro.

* 1x5 = 5
* 2x6 = 12
* 3x7 = 21
* 4x8 = 32

También podemos multiplicar un array por un número:

In [49]:
arr_1 * 2

array([2, 4, 6, 8])

En este casi sería equivalente a un array con una fila y una columna 1x1:

In [50]:
arr_4 = np.array(2)
arr_1 * arr_4

array([2, 4, 6, 8])

### División

In [51]:
arr_1 / arr_2

array([ 0.2       ,  0.33333333,  0.42857143,  0.5       ])

Igual que el producto, la división entre arrays se basa en dividir cada elemento de un array por el elemento en la misma posición del otro.

* 1/5 = 0.2
* 2/6 = 0.33...
* 3/7 = 0.42...
* 4/8 = 0.5

También podemos dividir todos sus elementos por un número:

In [52]:
arr_1 / 2

array([ 0.5,  1. ,  1.5,  2. ])

* 1/2 = 0.5
* 2/2 = 1
* 3/2 = 1.5
* 4/2 = 2

Algo interesante que podemos hacer con la división es conseguir el arreglo inverso o recíproco dividiendo 1 entre el array:

In [53]:
1 / arr_1

array([ 1.        ,  0.5       ,  0.33333333,  0.25      ])

Es equivalente hacer la potencia a -1 del array:

In [54]:
arr_1 ** -1

array([1, 0, 0, 0], dtype=int32)

Aunque como véis no da lo mismo, eso es porque al poner -1 estamos indicando la potencia a un número entero, y por tanto Numpy presupone que queremos el array resultante de tipo entero y nos redondea los valores.

Podemos simplemente indicar la elevación a -1. e indicar así que queremos decimales:

In [55]:
arr_1 ** -1.

array([ 1.        ,  0.5       ,  0.33333333,  0.25      ])

Ya que estamos con las potencias, también podemos hacer potencias entre arrays.

### Potencia

In [56]:
arr_1 ** arr_2

array([    1,    64,  2187, 65536], dtype=int32)

Como es normal, se basa en realizar la potencia entre los valores que comparten posición en los arrays:

* 1**5 = 5
* 2**6 = 64
* 3**7 = 2187
* 4**8 = 65536

### Operaciones en arrays de 2D
Todo lo que hemos visto aplica también a los arrays de dos dimensiones:

In [63]:
arr_5 = np.array([[1,2],[3,4]])
arr_6 = np.array([[5,6],[7,8]])

arr_5 + arr_6

array([[ 6,  8],
       [10, 12]])

Como es normal se basa en realizar la operación entre los valores que comparten posición en los arrays:

**[ [1+5, 2+6], [3+7, 4+8] ] = [ [6, 8], [10, 12] ] **

De igual forma funcionaría el producto, división y potencia por un número:

In [60]:
arr_5 * 3

array([[ 3,  6],
       [ 9, 12]])

**[ [1x3, 2x3], [3x3, 4x3] ] = [ [3, 6], [9, 12] ] **

Como hemos comentado antes **podemos multiplicar, dividir y potenciar matrices siempre que el número de columnas de la primera coincida con el número de filas de la segunda**.

In [66]:
arr_7 = np.array([5,10])

arr_5 * arr_7

array([[ 5, 20],
       [15, 40]])

**[ [1x5, 2x5], [3x10, 4x10] ] = [ [5, 10], [30, 40] ] **

¡Tanto estudiar álgebra de matrices para que al final un programa lo haga automáticamente! 

## Ejercicios